src/LanguageServer.Common/Utilities/DotnetInfo.cs

```diff
@@ -0,0 +1,835 @@
+
+            var dotnetRuntimes = new List<DotnetRuntimeInfo>();
+
+            string currentLine;
+            while ((currentLine = dotNetListRuntimesOutput.ReadLine()) != null)
+            {
+                Match parseResult = RuntimeInfoParser().Match(currentLine);
+                if (!parseResult.Success)
+                    continue;
+
+                if (!SemanticVersion.TryParse(parseResult.Groups["RuntimeVersion"].Value, out SemanticVersion runtimeVersion))
+                    continue;
+
+                string runtimeName = parseResult.Groups["RuntimeName"].Value;
+
+                dotnetRuntimes.Add(
+                    new DotnetRuntimeInfo(runtimeName, runtimeVersion)
+                );
+            }
+
+            return dotnetRuntimes;
+        }
+
+        /// <summary>
+        ///     Invoke the .NET host ("dotnet").
+        /// </summary>
+        /// <param name="commandLineArguments">
+        ///     Command-line arguments to be passed to the host.
+        /// </param>
+        /// <param name="baseDirectory">
+        ///     The directory where the host will be invoked.
+        /// </param>
+        /// <param name="logger">
+        ///     The logger for diagnostic messages.
+        /// </param>
+        /// <param name="enableHostTracing">
+        ///     Enable host-level tracing?
+        /// </param>
+        /// <returns>
+        ///     The process exit code, and a <see cref="TextReader"/> containing the program output (STDOUT and STDERR).
+        /// </returns>
+        static (DotNetHostExitCode ExitCode, TextReader StdOut) InvokeDotNetHost(string commandLineArguments, string baseDirectory, ILogger logger, bool enableHostTracing = false)
+        {
+            if (logger == null)
+                throw new ArgumentNullException(nameof(logger));
+
+            var dotnetHostProcess = new Process
+            {
+                StartInfo = new ProcessStartInfo
+                {
+                    FileName = Environment.GetEnvironmentVariable("DOTNET_HOST_PATH") ?? "dotnet",
+                    WorkingDirectory = baseDirectory,
+                    Arguments = commandLineArguments,
+                    UseShellExecute = false,
+                    RedirectStandardOutput = true,
+                    RedirectStandardError = true,
+                    Environment =
+                    {
+                        ["COREHOST_TRACE"] = enableHostTracing ? "1" : "0",
+                    }
+                },
+                EnableRaisingEvents = true
+            };
+
+            using (dotnetHostProcess)
+            {
+                // For logging purposes.
+                string command = $"{dotnetHostProcess.StartInfo.FileName} {dotnetHostProcess.StartInfo.Arguments}";
+
+                // Buffer the output locally (otherwise, the process may hang if it fills up its STDOUT / STDERR buffer).
+                var stdOutBuffer = new StringBuilder();
```

Comment:
We should probably use a library for invoking external commands to avoid all this complications. Not in this PR of course